llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -21834,15 +21824,23 @@ SDValue tryLowerPartialReductionToWideAdd(SDNode *N,
   auto Acc = N->getOperand(0);
   auto Input = N->getOperand(1);
 
-  EVT AccVT = Acc.getValueType();
+  unsigned Opcode = N->getOpcode();
+  unsigned InputOpcode = Input.getOpcode();
+  if (ISD::isExtOpcode(InputOpcode)) {
+    Input = Input.getOperand(0);
+    if (InputOpcode == ISD::SIGN_EXTEND)
+      Opcode = ISD::PARTIAL_REDUCE_SADD;
+  }
+
```

Comment:
Currently there would be no node for `ISD::PARTIAL_REDUCE_SADD` made anymore. But with the legalisation, I would make it output a node with the opcode `ISD::PARTIAL_REDUCE_SADD` here. This would then go through legalisation and the execution would then lead back to `tryLowerPartialReductionToWideAdd`. 
I originally left this out of this patch because it wasn't necessary to get it to work. But I will go back and make it return the node, just to make sure it all works properly.

Reply:
I've now made it so the DAG-combine function decides whether to keep the node opcode as `ISD::PARTIAL_REDUCE_UADD` or whether to convert it to `ISD::PARTIAL_REDUCE_SADD`. But, the DAG-combine function also includes the lowering for `usdot` IR patterns (due to the reasons described in the above comment). Also, the DAG-combine needed to include lowering for `(nx)v16i8` to `(nx)v4i64`, because otherwise the instruction would be attempted to be legalised. But, there is now a lowering function called by `LowerOperation` which covers the general cases for lowering to wide add / dot product instructions.